torch/distributed/fsdp/ssd_offload.py

```diff
@@ -0,0 +1,958 @@
+            module (nn.Module): module into which to inject the property
+            property_name (str): name of the name of the property to create
+        """
+
+        def get_propertized(self: torch.nn.Module) -> torch.Tensor:
+            prop: Callable[[], torch.Tensor] = self.__override_properties__[property_name]  # type: ignore
+            # If caching is not active, this function just evaluates the propertization
+            return prop()
+
+        def set_original(self: torch.nn.Module, value: Callable[[], torch.Tensor]) -> None:
+            self.__override_properties__[property_name] = value  # type: ignore
+
+        def del_fn(self: torch.nn.Module) -> None:
+            _remove_property(self, property_name)
+
+        setattr(module.__class__, property_name, property(get_propertized, set_original, del_fn))
+
+
+    @staticmethod
+    def register_property(module: torch.nn.Module, property_name: str, property_value: Callable[[], torch.Tensor]) -> None:
+        has_injected_class = hasattr(module, "__override_properties__")
+        if not has_injected_class:
+            PropertizeModule.__inject_new_class(module)
+            if hasattr(module, property_name):
+                delattr(module, property_name)
+        module.__override_properties__[property_name] = property_value  # type: ignore
+        PropertizeModule.__inject_property(module, property_name)
+
+
+    @staticmethod
+    def remove_property(module: torch.nn.Module, property_name: str, new_property_value: Optional[Any] = None) -> None:
+        delattr(module.__class__, property_name)
+        del module.__override_properties__[property_name]  # type: ignore
+
+        # Roll back the parametrized class if no other buffer or parameter
+        # is currently parametrized in this class
+        if len(module.__override_properties__) == 0:  # type: ignore
+            delattr(module, "__override_properties__")
+            # Restore class
+            orig_cls = module.__class__.__bases__[0]
+            module.__class__ = orig_cls
+        if new_property_value is not None:
+            setattr(module.__class__, property_name, new_property_value)
+
+
+class SsdFlatParameterViewProperty:
+    """
+    Allows for a mutable view to replace a layer's trainable parameters.
+    This is needed since FSDP is changing tensor metadata (.data) under the covers,
+    SsdFlatParameter cannot just rely on this since each view (of type SsdFlatParameterView) has
+    an internal representation. So every time we access a view, we need to
+    make sure we get the up-to-date(with the updated .data info) version, and not
+     the original version.
+    """
+
+    def __init__(self, parent: SsdFlatParameter, view_id: int) -> None:
+        super().__init__()
+        self.parent = parent
+        self.view_id = view_id
+
+    def __call__(self) -> SsdFlatParameterView:
+        return self.parent.views[self.view_id]
+
+# Classes supporting torch.save/load
+class TorchSaver:
```

Comment:
I guess for SSD offload use cases, we may have to save model shards, right now we save model shards as shardedTensor, shardedTensor uses its own saving and loading APIs, could we move this change into shardedTensor's saving and loading APIs, so that the checkpointing saving and loading APIs are unified

Reply:
oh sorry, shardedTensor can be used with torch.save and torch.load as well

Reply:
I'm not quite sure I understand what your suggestion is here.